mava/wrappers/gigastep.py

```diff
@@ -0,0 +1,332 @@
+        """
+        Takes a step in the Gigastepl environment.
+
+        Args:
+            state (GigasteplState): The current state of the environment.
+            action (Array): The actions for controllable agents.
+
+        Returns:
+            Tuple[GigasteplState, TimeStep]: A tuple containing the next state of the environment
+            and the next time step.
+
+        """
+        key, step_key, adversary_key = jax.random.split(state.key, 3)
+
+        action = jnp.concatenate([action, state.Adverseary_action], axis=0, dtype=jnp.int16)
+
+        obs, env_state, rewards, dones, ep_done = self._env.step(state.state, action, step_key)
+
+        rewards, dones = (
+            rewards[: self.num_agents],
+            dones[: self.num_agents],
+        )  # cut out the rewards,dones of the adversary
+        team1_obs, team1_state, team2_obs, team2_state = self._split_obs_and_state(obs, env_state)
+        adversary_actions = self.adversary_policy(
+            team2_obs, team2_state, adversary_key
+        )  # take the actions of the adversary and cache it befor returning the new state
+
+        if self.has_global_state:
+            obs = ObservationGlobalState(
+                agents_view=team1_obs,
+                action_mask=self.action_mask(),
+                global_state=self.get_global_state(obs, adversary_actions),
+                step_count=jnp.zeros(self.num_agents, dtype=int),
+            )
+        else:
+            obs = Observation(
+                agents_view=team1_obs,
+                action_mask=self.action_mask(),
+                step_count=jnp.zeros(self.num_agents, dtype=int),
+            )
+
+        step_type = jax.lax.select(ep_done, StepType.LAST, StepType.MID)
+
+        current_winner = jnp.where(ep_done, self.get_wining_team(env_state), jnp.nan)
+
+        ts = TimeStep(
+            step_type=step_type,
+            reward=rewards,
+            discount=1.0 - dones,
+            observation=obs,
+            extras={"won_episode": current_winner},
+        )
+        return GigasteplState(env_state, key, state.step + 1, adversary_actions), ts
+
+    def action_mask(self) -> Array:
+        """Get action mask for each agent."""
+        return jnp.ones((self.num_agents, self._env.n_actions))  # all actions are valid
+
+    def get_global_state(self, obs: Array, adversary_actions: Array) -> Array:
+        """
+        Combines observations from all agents,
+        optionally adding adversary actions, and adversary observations
+        to create a global state for the environment.
+
+        Args:
+            obs (Array): The observations of all agents.
+            adversary_actions (Array): The actions for the adversary.
+
+        Returns:
+            global_obs (Array): The global observation.
+        """
+        global_obs = jnp.concatenate(
+            obs if self.get_adversary_obs else obs[: self.num_agents], axis=0
+        )
+        if self.get_adversary_actions:
```

Comment:
I don't think this should be optional, we should either always or never add it.

Unless they use this in the paper I would say just leave it out

Reply:
Yes I would say also leave it for, they didn't use it for their [baselines](https://github.com/mlech26l/gigastep/blob/6288d650b1697ae6b201dcc7db9cbe0e07de37a2/baselines/bline_ppo.py#L250C37-L250C45).